odoo/models.py

```diff
@@ -3151,7 +3151,7 @@ def _check_company(self, fnames=None):
                     if not (company <= record[name].company_ids):
                         inconsistent_fields.add(name)
                         inconsistent_recs |= record
-                elif not (record[name].company_id <= company):
+                elif not (record.sudo()[name].company_id <= company):
```

Comment:
The cases above (`record[name].company_ids`) are not protected.
I suggest to factor out `record[name]` like
```python
corecord = record.sudo()[name]
```
and use `corecord` in the 4 places.